### PR TITLE
[Issue 15896] Fix LockTimeout when storePut on the same key concurrently in RocksdbMetadataStore

### LINKED ISSUE #15896: Flaky-test: PulsarLedgerIdGeneratorTest.testGenerateLedgerId
org.apache.pulsar.metadata.bookkeeper.PulsarLedgerIdGeneratorTest is flaky. The testGenerateLedgerId test method fails sporadically.

```
java.lang.AssertionError: Error occur during ledger id generation :  expected [0] but found [2]
	at org.testng.Assert.fail(Assert.java:99)
	at org.testng.Assert.failNotEquals(Assert.java:1037)
	at org.testng.Assert.assertEqualsImpl(Assert.java:140)
	at org.testng.Assert.assertEquals(Assert.java:122)
	at org.testng.Assert.assertEquals(Assert.java:907)
	at org.apache.pulsar.metadata.bookkeeper.PulsarLedgerIdGeneratorTest.testGenerateLedgerId(PulsarLedgerIdGeneratorTest.java:113)
```

Usage tip: To enable automatic navigation to failure message, open the following links with CTRL/CMD-click.  
[example failure 2022-06-01T14:25:28.8616198Z](https://github.com/apache/pulsar/runs/6691176380?check_suite_focus=true#step:9:1542)  
[example failure 2022-05-27T11:16:52.6256505Z](https://github.com/apache/pulsar/runs/6623892528?check_suite_focus=true#step:9:2931)  
[example failure 2022-05-27T11:13:47.4439338Z](https://github.com/apache/pulsar/runs/6623892528?check_suite_focus=true#step:9:1534)  

<details>
<summary>Full exception stacktrace</summary>
<code><pre>
java.lang.AssertionError: Error occur during ledger id generation :  expected [0] but found [2]
	at org.testng.Assert.fail(Assert.java:99)
	at org.testng.Assert.failNotEquals(Assert.java:1037)
	at org.testng.Assert.assertEqualsImpl(Assert.java:140)
	at org.testng.Assert.assertEquals(Assert.java:122)
	at org.testng.Assert.assertEquals(Assert.java:907)
	at org.apache.pulsar.metadata.bookkeeper.PulsarLedgerIdGeneratorTest.testGenerateLedgerId(PulsarLedgerIdGeneratorTest.java:113)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:132)
	at org.testng.internal.InvokeMethodRunnable.runOne(InvokeMethodRunnable.java:45)
	at org.testng.internal.InvokeMethodRunnable.call(InvokeMethodRunnable.java:73)
	at org.testng.internal.InvokeMethodRunnable.call(InvokeMethodRunnable.java:11)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:833)

</pre></code>
</details>

```
java.lang.AssertionError: Error occur during ledger id generation :  expected [0] but found [12]
	at org.testng.Assert.fail(Assert.java:99)
	at org.testng.Assert.failNotEquals(Assert.java:1037)
	at org.testng.Assert.assertEqualsImpl(Assert.java:140)
	at org.testng.Assert.assertEquals(Assert.java:122)
	at org.testng.Assert.assertEquals(Assert.java:907)
	at org.apache.pulsar.metadata.bookkeeper.PulsarLedgerIdGeneratorTest.testGenerateLedgerId(PulsarLedgerIdGeneratorTest.java:113)
```

Usage tip: To enable automatic navigation to failure message, open the following links with CTRL/CMD-click.  
[example failure 2022-05-30T22:43:38.9123608Z](https://github.com/apache/pulsar/runs/6661894942?check_suite_focus=true#step:10:1517)  

<details>
<summary>Full exception stacktrace</summary>
<code><pre>
java.lang.AssertionError: Error occur during ledger id generation :  expected [0] but found [12]
	at org.testng.Assert.fail(Assert.java:99)
	at org.testng.Assert.failNotEquals(Assert.java:1037)
	at org.testng.Assert.assertEqualsImpl(Assert.java:140)
	at org.testng.Assert.assertEquals(Assert.java:122)
	at org.testng.Assert.assertEquals(Assert.java:907)
	at org.apache.pulsar.metadata.bookkeeper.PulsarLedgerIdGeneratorTest.testGenerateLedgerId(PulsarLedgerIdGeneratorTest.java:113)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:132)
	at org.testng.internal.InvokeMethodRunnable.runOne(InvokeMethodRunnable.java:45)
	at org.testng.internal.InvokeMethodRunnable.call(InvokeMethodRunnable.java:73)
	at org.testng.internal.InvokeMethodRunnable.call(InvokeMethodRunnable.java:11)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:833)

</pre></code>
</details>

```
java.lang.AssertionError: Error occur during ledger id generation :  expected [0] but found [3]
	at org.testng.Assert.fail(Assert.java:99)
	at org.testng.Assert.failNotEquals(Assert.java:1037)
	at org.testng.Assert.assertEqualsImpl(Assert.java:140)
	at org.testng.Assert.assertEquals(Assert.java:122)
	at org.testng.Assert.assertEquals(Assert.java:907)
	at org.apache.pulsar.metadata.bookkeeper.PulsarLedgerIdGeneratorTest.testGenerateLedgerId(PulsarLedgerIdGeneratorTest.java:113)
```

Usage tip: To enable automatic navigation to failure message, open the following links with CTRL/CMD-click.  
[example failure 2022-05-31T11:13:55.0547600Z](https://github.com/apache/pulsar/runs/6669032761?check_suite_focus=true#step:9:4932)  

<details>
<summary>Full exception stacktrace</summary>
<code><pre>
java.lang.AssertionError: Error occur during ledger id generation :  expected [0] but found [3]
	at org.testng.Assert.fail(Assert.java:99)
	at org.testng.Assert.failNotEquals(Assert.java:1037)
	at org.testng.Assert.assertEqualsImpl(Assert.java:140)
	at org.testng.Assert.assertEquals(Assert.java:122)
	at org.testng.Assert.assertEquals(Assert.java:907)
	at org.apache.pulsar.metadata.bookkeeper.PulsarLedgerIdGeneratorTest.testGenerateLedgerId(PulsarLedgerIdGeneratorTest.java:113)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:132)
	at org.testng.internal.InvokeMethodRunnable.runOne(InvokeMethodRunnable.java:45)
	at org.testng.internal.InvokeMethodRunnable.call(InvokeMethodRunnable.java:73)
	at org.testng.internal.InvokeMethodRunnable.call(InvokeMethodRunnable.java:11)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:833)

</pre></code>
</details>

### PR DESCRIPTION
**Pull request by Jason918**
Fixes #15896

### Motivation

See #15896
When the same key is modified concurrently, there will be LockTimeout exception in `RocksdbMetadataStore`.

The root cause is we haven't get the lock of the key when read old data in `store.getForUpdate(...)`, so the exception throws when two transaction modifies the same key.

### Modifications

Set `exclusive` parameter of `getForUpdate` to `true` in `RocksdbMetadataStore#storePut`.

The same applies to  `RocksdbMetadataStore#storeDelete`.

### Verifying this change

- [ ] Make sure that the change passes the CI checks.

This change added tests and can be verified as follows:

  - org.apache.pulsar.metadata.MetadataStoreTest#testConcurrentPut
  - org.apache.pulsar.metadata.MetadataStoreTest#testConcurrentDelete
### Does this pull request potentially affect one of the following parts:

*If `yes` was chosen, please highlight the changes*

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API: (no)
  - The schema: (no)
  - The default values of configurations: (no)
  - The wire protocol: (no)
  - The rest endpoints: (no)
  - The admin cli options: (no)
  - Anything that affects deployment: (no)

### Documentation

Check the box below or label this PR directly.

Need to update docs?

 - [x] doc-not-needed (bug fix)

### CODE REVIEW

File: pulsar-metadata/src/test/java/org/apache/pulsar/metadata/MetadataStoreTest.java [hunk 1]
```diff
@@ -447,4 +448,33 @@ public void run() {
         assertEquals(successWrites.get(), maxValue);
         assertEquals(store.get(path).get().get().getValue()[0], maxValue);
     }
+
+    @Test(dataProvider = "impl")
+    public void testConcurrentPut(String provider, Supplier<String> urlSupplier) throws Exception {
+        MetadataStore store = MetadataStoreFactory.create(urlSupplier.get(), MetadataStoreConfig.builder().build());
```
Comment (codelipenghui) [inline pulsar-metadata/src/test/java/org/apache/pulsar/metadata/MetadataStoreTest.java:454]: Need to close the store after test?

Comment (Jason918) [reply pulsar-metadata/src/test/java/org/apache/pulsar/metadata/MetadataStoreTest.java:455]: Yep, added.

File: pulsar-metadata/src/test/java/org/apache/pulsar/metadata/MetadataStoreTest.java [hunk 2]
```diff
@@ -447,4 +448,33 @@ public void run() {
...
+
+        String k = newKey();
+        CompletableFuture<Void> f1 =
+                CompletableFuture.runAsync(() -> store.put(k, new byte[0], Optional.of(-1L)).join());
+        CompletableFuture<Void> f2 =
+                CompletableFuture.runAsync(() -> store.put(k, new byte[0], Optional.of(-1L)).join());
```
[the remark this answers is not in the document]
Comment (Jason918) [reply pulsar-metadata/src/test/java/org/apache/pulsar/metadata/MetadataStoreTest.java:461]: Do you mean increase concurrency? 
It's not necessary in this case, two threads can reproduce the issue easily.

**Review by RobertIndie — APPROVED** [review]

**Comment by Jason918** [comment]
@codelipenghui @wuxuanqicn PTAL

**Comment by Jason918** [comment]
@merlimat PTAL

**Review by merlimat — APPROVED** [review]

**Pull request merged**